R/Vietnam.R

```diff
@@ -0,0 +1,129 @@
+                         supported_region_codes = list("1" = "iso_3166_2"),
+                         #' @field common_data_urls List of named links to raw data.
+                         common_data_urls = list(),
+                         #' @field source_data_cols existing columns within the raw data
+                         source_data_cols = c(
+                           "cases_total", "deaths_total", "recovered_total"
+                         ),
+                         #' @field source_text Plain text description of the source of the data
+                         source_text = "Public COVID-19 for Vietnam, curated by NCSC's COVID-19 team",
+                         #' @field source_url Website address for explanation/introduction of the
+                         #' data
+                         source_url = "https://covid.ncsc.gov.vn", # nolint
+                         
+                         #' @description Set up a table of region codes for clean data
+                         #' @importFrom tibble tibble
+                         set_region_codes = function(){
+                           self$codes_lookup$`1` <- covidregionaldata::vietnam_codes
+                         },
+                         
+                         #' @description download function to get raw data
+                         #' @importFrom tidyr replace_na drop_na
+                         #' @importFrom lubridate dmy
+                         #' @importFrom jsonlite fromJSON
+                         download = function() {
+                           bundles_urls = list(
+                             "case_by_time" = 'https://covid.ncsc.gov.vn/api/v3/covid/provinces?filter_type=case_by_time',
+                             "death_by_time" = 'https://covid.ncsc.gov.vn/api/v3/covid/provinces?filter_type=death_by_time',
+                             "recovered_by_time" = 'https://covid.ncsc.gov.vn/api/v3/covid/provinces?filter_type=recovered_by_time'
+                             
+                           ) 
+                           Sys.setenv("VROOM_CONNECTION_SIZE" = 131072*4) # Fix VROOM error
+                           provines_url = 'https://covid.ncsc.gov.vn/api/v3/covid/provinces'
+                           bundles = names(bundles_urls)
+                           provines_data = fromJSON(provines_url)
+                           
+                           get_bundles_data = function(bundles){
+                             bundles_data = list()
+                             for (bundle in bundles){
+                               url = paste0('https://covid.ncsc.gov.vn/api/v3/covid/provinces?filter_type=', bundle)
+                               data = fromJSON(url)
+                               bundles_data = c(bundles_data, setNames(list(data), bundle))
+                             }
+                             bundles_data
+                           }
+                           
```

Comment:
This feels like it's similar to what the standard `download` function does with `purl` and could probably be rewritten to emulate that structure a bit. We're building `bundles_data[case_by_time]`, `bundles_data[death_by_time]` and `bundles_data[recovered_by_time]`. It would be more straightforward to put these into `self$data$raw[main]`, `self$data$raw[death_by_time]` and `self$data$raw[recovered_by_time]` ... [see next comment]